specs/MultiProfile.md

```diff
@@ -0,0 +1,240 @@
+      [out, retval] ICoreWebView2ControllerOptions** options);
+
+  /// Create a new WebView with options.
+  HRESULT CreateCoreWebView2ControllerWithOptions(
+      [in] HWND parentWindow,
+      [in] ICoreWebView2ControllerOptions* options,
+      [in] ICoreWebView2CreateCoreWebView2ControllerCompletedHandler* handler);
+
+  /// Create a new WebView in visual hosting mode with options.
+  HRESULT CreateCoreWebView2CompositionControllerWithOptions(
+      [in] HWND parentWindow,
+      [in] ICoreWebView2ControllerOptions* options,
+      [in] ICoreWebView2CreateCoreWebView2CompositionControllerCompletedHandler* handler);
+}
+
+[uuid(6E5CE5F0-16E6-4A05-97D8-4E256B3EB609), object, pointer_default(unique)]
+interface ICoreWebView2_7 : IUnknown {
+  /// The associated `ICoreWebView2Profile` object.
+  [propget] HRESULT Profile([out, retval] ICoreWebView2Profile** profile);
+}
+
+[uuid(3B9A2AF2-E703-4C81-9D25-FCE44312E960), object, pointer_default(unique)]
+interface ICoreWebView2Profile : IUnknown {
+  /// Name of the profile.
+  [propget] HRESULT ProfileName([out, retval] LPWSTR* value);
+
+  /// InPrivate mode is enabled or not.
+  [propget] HRESULT InPrivateModeEnabled([out, retval] BOOL* enabled);
+
+  /// Full path of the profile directory.
+  [propget] HRESULT ProfilePath([out, retval] LPWSTR* value);
+
+  // TODO: All profile-wide operations/settings will be put below in the future.
+}
+```
+
+## .NET and WinRT
+
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    runtimeclass CoreWebView2ControllerOptions;
+    runtimeclass CoreWebView2Environment;
+    runtimeclass CoreWebView2;
+    runtimeclass CoreWebView2Profile;
+
+    runtimeclass CoreWebView2ControllerOptions
+    {        
+        String ProfileName { get; set; };
+
+        Int32 InPrivateModeEnabled { get; set; };
+    }
+    
+    runtimeclass CoreWebView2Environment
+    {
+        // ...
+    
+        CoreWebView2ControllerOptions CreateCoreWebView2ControllerOptions(
+            String ProfileName, Int32 InPrivateModeEnabled);
+        
+        Windows.Foundation.IAsyncOperation<CoreWebView2Controller>
+        CreateCoreWebView2ControllerWithOptionsAsync(
+            CoreWebView2ControllerWindowReference ParentWindow,
+            CoreWebView2ControllerOptions options);
+        
+        Windows.Foundation.IAsyncOperation<CoreWebView2CompositionController>
+        CreateCoreWebView2CompositionControllerWithOptionsAsync(
+            CoreWebView2ControllerWindowReference ParentWindow,
+            CoreWebView2ControllerOptions options);
+    }
+    
+    runtimeclass CoreWebView2
+    {
+        // ...
+        CoreWebView2Profile Profile { get; };
+    }
+    
+    runtimeclass CoreWebView2Profile
+    {
+        String ProfileName { get; };
+
+        Int32 InPrivateModeEnabled { get; };
+
+        String ProfilePath { get; };
+    }
+}
+```
+
+# Appendix
+
+Next we'll consolidate all profile-wide operations/settings into the interface
```

Comment:
Do we already know what the APIs should look like to consolidate existing profile-wide operations here and to add a method to erase a profile? Can you please describe the future changes with slightly more detail in this issue?

I just want to make sure that we don't expect those future changes to make us regret the decisions we're making in this document now. We could potentially do a larger API review document that includes those changes if we think those future additions will influence our thoughts on the current APIs. I'm guessing that's not necessary but just to be safe please list the future changes.

Thank you!

Reply:
Added more details for the future changes. The erase method might need relatively complex implementation, so maybe we do not want to implement it unless strong requests come up. Other methods are supposed to be straightforward.

Reply:
Thanks! I meant just in the issue here. Please remove the IDL section below from the doc which may be confusing for API reviewers.

Reply:
Removed from this doc.

The idea for future changes is roughly illustrated as below.

## Win32 C++

```IDL
[uuid(3B9A2AF2-E703-4C81-9D25-FCE44312E960), object, pointer_default(unique)]
interface ICoreWebView2Profile : IUnknown {
  // ...
  
  /// Gets the cookie manager object for this profile.
  /// See ICoreWebView2CookieManager.
  [propget] HRESULT CookieManager([out, retval] ICoreWebView2CookieManager** cookieManager);

  /// Clears browsing data based on a specific data type. The completed handler
  /// will be invoked when the browsing data has been cleared and will indicate
  /// if the specified data was properly cleared.
  HRESULT ClearBrowsingData(
      [in] COREWEBVIEW2_BROWSING_DATA_KIND dataKind,
      [in] ICoreWebView2ClearBrowsingDataCompletedHandler *handler);

  /// Toggles whether autosave for password information is enabled. Disabled by default.
  HRESULT TogglePasswordAutosaveEnabled([in] BOOL enable);

  /// Toggles whether autofill for information like names, street and email addresses, phone
  /// numbers, and arbitrary input is enabled. This excludes password and credit card
  /// information. Enabled by default. 
  HRESULT ToggleGeneralAutofillEnabled([in] BOOL enable);

  /// Erase the profile completely. All webviews on this profile will be closed and the profile
  /// directory on disk will be deleted.
  HRESULT Erase([in] ICoreWebView2ProfileEraseCompletedHandler *handler);
}
```